src/main/java/com/aws/greengrass/deployment/DefaultDeploymentTask.java

```diff
@@ -194,6 +195,18 @@ public DeploymentResult call() throws InterruptedException {
         Optional<Set<String>> groupsForDeviceOpt;
         try {
             groupsForDeviceOpt = thingGroupHelper.listThingGroupsForDevice(retryCount);
+        } catch (GreengrassV2DataException e) {
+            if (e.statusCode() == 403) {
+                // Getting group hierarchy requires permission to call the ListThingGroupsForCoreDevice API which
+                // may not be configured on existing IoT Thing policy in use for current device, log a warning in
+                // that case and move on.
+                logger.atWarn().setCause(e).log("Failed to get thing group hierarchy. Deployment will proceed. "
+                        + "To automatically clean up unused components, please add "
+                        + "greengrass:ListThingGroupsForCoreDevice permission to your IoT Thing policy.");
+                groupsForDeviceOpt = getPersistedMembershipInfo();
+            } else {
+                throw new DeploymentTaskFailureException("Error fetching thing group information", e);
+            }
         } catch (Exception e) {
             if (isLocalDeployment && ThingGroupHelper.DEVICE_OFFLINE_INDICATIVE_EXCEPTIONS.contains(e.getClass())) {
                 logger.atWarn().setCause(e).log("Failed to get thing group hierarchy, local deployment will proceed");
```

Comment:
Not blocking / follow-up checkin: This feels wrong - device may go offline, doesn't mean it's a local deployment.

Reply:
We know this is a local deployment at this point. There is a check above for isLocalDeployment

Reply:
I see it now, yes.